lab_2_tokenize_by_bpe/main.py

```diff
@@ -26,42 +38,88 @@ def collect_frequencies(
     :param end_of_word: a token that signifies the end of word
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not (isinstance(text, str) and isinstance(end_of_word, str)) or not (
+            isinstance(start_of_word, str) or start_of_word is None
+    ):
+        return None
+    freq_dict = {}
+    words = text.split()
+    for word in words:
+        preprocessed_word = prepare_word(word, start_of_word, end_of_word)
+        if not preprocessed_word:
+            return None
+        freq_dict.update({preprocessed_word: words.count(word)})
+    return freq_dict
 
 
 def count_tokens_pairs(
-    word_frequencies: dict[tuple[str, ...], int]
+        word_frequencies: dict[tuple[str, ...], int]
 ) -> dict[tuple[str, str], int] | None:
     """
     Counts number of occurrences of each pair of subsequent tokens
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+    pairs_of_tokens = {}
+    for word, freq in word_frequencies.items():
+        for ind in range(len(word) - 1):
+            token_pair = (word[ind], word[ind + 1])
+            if token_pair not in pairs_of_tokens:
+                pairs_of_tokens[token_pair] = freq
+            pairs_of_tokens[token_pair] += freq
+    return pairs_of_tokens
 
 
 def merge_tokens(
-    word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
+        word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
 ) -> dict[tuple[str, ...], int] | None:
     """
     Updates word frequency dictionary by replacing a pair of token with a merged one
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(pair, tuple):
+        return None
+    merged_dict = {}
+    for word, freq in word_frequencies.items():
+        new_word = ''.join(word).replace(pair[0] + '' + pair[1], pair[0] + pair[1])
+        new_word_tupled = tuple(new_word.split())
+        if new_word_tupled not in merged_dict:
+            merged_dict[new_word_tupled] += freq
+        else:
```

Comment:
Try to avoid `else` here, too

Reply:
I've tried to change the code here. Could you please check and let me know whether it's possible and appropriate the way i did it?